File: src/main/java/in/tazj/k8s/letsencrypt/kubernetes/SecretManager.java
package in.tazj.k8s.letsencrypt.kubernetes;

import com.google.common.collect.ImmutableMap;

import org.joda.time.LocalDate;

import java.util.Optional;

import in.tazj.k8s.letsencrypt.model.CertificateRequest;
import in.tazj.k8s.letsencrypt.model.CertificateResponse;
import io.fabric8.kubernetes.api.model.Secret;
import io.fabric8.kubernetes.client.KubernetesClient;
import lombok.extern.slf4j.Slf4j;
import lombok.val;

import static in.tazj.k8s.letsencrypt.model.Constants.ACME_CA_ANNOTATION;
import static in.tazj.k8s.letsencrypt.model.Constants.EXPIRY_ANNOTATION;
import static in.tazj.k8s.letsencrypt.model.Constants.REQUEST_ANNOTATION;

/**
 * Manages certificates in the form of secrets in the Kubernetes API.
 */
@Slf4j
public class SecretManager {
  final private KubernetesClient client;

  public SecretManager(KubernetesClient client) {
    this.client = client;
  }

  public Optional<Secret> getSecret(String namespace, String secretName) {
    // TODO: .get() returns null but .list() works, not sure why yet.
    final Optional<Secret> secret = client.secrets().inNamespace(namespace)
        .list().getItems().stream()
        .filter(s -> s.getMetadata().getName().equals(secretName))
        .findFirst();
    return secret;
  }

  /**
   * Insert a specified certificate in the Kubernetes cluster.
   */
  public void insertCertificate(String namespace, String secretName,
                                CertificateResponse certificate) {
    val expiryDate = LocalDate.fromDateFields(certificate.getExpiryDate());
    val annotations = ImmutableMap.of(
        EXPIRY_ANNOTATION, expiryDate.toString(),
        ACME_CA_ANNOTATION, certificate.getCa());

    client.secrets().inNamespace(namespace)
        .createNew()
        .withNewMetadata()
          .withName(secretName)
          .withAnnotations(annotations)
        .endMetadata()
        .withData(certificate.getCertificateFiles())
        .done();

    log.info("Inserted secret {} into namespace {}", secretName, namespace);
  }

  public void updateCertificate(String namespace,
                                String secretName,
                                CertificateResponse certificate) {
    val expiryDate = LocalDate.fromDateFields(certificate.getExpiryDate());

    client.secrets().inNamespace(namespace).withName(secretName).edit()
        .editMetadata()
          .removeFromAnnotations(EXPIRY_ANNOTATION)
          .addToAnnotations(EXPIRY_ANNOTATION, expiryDate.toString())
          .addToAnnotations(ACME_CA_ANNOTATION, certificate.getCa())
        .endMetadata()
        .withData(certificate.getCertificateFiles())
        .done();

    log.info("Updated secret {} in namespace {}", secretName, namespace);
  }

  /**
   * Checks whether a certificate needs renewal. Renewal can be caused either by certificate expiry
   * or if the list of domains requested for a SAN certificate changes.
   */
  public static boolean certificateNeedsRenewal(String requestAnnotation,
                                                Secret secret) {
    val isExpiring = certificateIsExpiring(secret);
    val domainsChanged = haveDomainsChanged(requestAnnotation, secret);

    return (isExpiring || domainsChanged);
  }

  private static boolean haveDomainsChanged(String requestAnnotation, Secret secret) {
    val annotations = secret.getMetadata().getAnnotations();

    if (annotations != null ) {
      val secretDomains = annotations.get(REQUEST_ANNOTATION);

      if (secretDomains != null) {
        val domainsChanged = !(requestAnnotation.equals(secretDomains));
        return domainsChanged;
      }
    }

    log.warn("acme/certificate annotation missing on secret {}!", secret.getMetadata().getName());
    return false;
  }

  public static boolean certificateIsExpiring(Secret secret) {
    val expiryDate = getExpiryDate(secret);

    if (expiryDate.isPresent()) {
      return LocalDate.now().isAfter(expiryDate.get().minusDays(2));
    } else {
      log.warn("No expiry date set on secret {} in namespace {}!",
          secret.getMetadata().getName(), secret.getMetadata().getNamespace());
      return false;
    }
  }

  private static Optional<LocalDate> getExpiryDate(Secret secret) {
    val annotations = secret.getMetadata().getAnnotations();

    if (annotations != null && annotations.get(EXPIRY_ANNOTATION) != null) {
      val annotation = annotations.get(EXPIRY_ANNOTATION);
      return Optional.of(new LocalDate(annotation));
    } else {
      return Optional.empty();
    }
  }
}

File: src/main/kotlin/in/tazj/k8s/letsencrypt/kubernetes/SecretManager.kt
package `in`.tazj.k8s.letsencrypt.kubernetes

import `in`.tazj.k8s.letsencrypt.model.ACME_CA_ANNOTATION
import `in`.tazj.k8s.letsencrypt.model.CertificateResponse
import `in`.tazj.k8s.letsencrypt.model.EXPIRY_ANNOTATION
import `in`.tazj.k8s.letsencrypt.model.REQUEST_ANNOTATION
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import io.fabric8.kubernetes.api.model.Secret
import io.fabric8.kubernetes.client.KubernetesClient
import org.funktionale.option.Option
import org.funktionale.option.getOrElse
import org.funktionale.option.toOption
import org.slf4j.LoggerFactory
import java.time.LocalDate
import java.time.ZoneId

/**
 * Manages certificates in the form of secrets in the Kubernetes API.
 */
class SecretManager(val client: KubernetesClient) {
    fun getSecret(namespace: String, secretName: String): Option<Secret> {
        // TODO: .get() returns null but .list() works, not sure why yet.
        return client.secrets().inNamespace(namespace)
                .list().items
                .filter { it.metadata.name == secretName }
                .lastOrNull().toOption()
    }

    /**
     * Insert a specified certificate in the Kubernetes cluster.
     */
    fun insertCertificate(namespace: String, secretName: String, certificate: CertificateResponse) {
        val expiryDate = certificate.expiryDate
                .toInstant()
                .atZone(ZoneId.systemDefault())
                .toLocalDate()
        val domains = gson.toJson(certificate.domains)
        val annotations = mapOf(
                EXPIRY_ANNOTATION to expiryDate.toString(),
                ACME_CA_ANNOTATION to certificate.ca,
                REQUEST_ANNOTATION to domains
        )

        client.secrets().inNamespace(namespace)
                .createNew()
                .withNewMetadata()
                .withName(secretName)
                .withAnnotations(annotations)
                .endMetadata()
                .withData(certificate.certificateFiles)
                .done()

        log.info("Inserted secret {} into namespace {}", secretName, namespace)
    }

    fun updateCertificate(namespace: String,
                          secretName: String,
                          certificate: CertificateResponse) {
        val expiryDate = certificate.expiryDate
                .toInstant()
                .atZone(ZoneId.systemDefault())
                .toLocalDate()

        val domains = gson.toJson(certificate.domains)

        client.secrets().inNamespace(namespace).withName(secretName).edit()
                .editMetadata()
                .removeFromAnnotations(EXPIRY_ANNOTATION)
                .addToAnnotations(EXPIRY_ANNOTATION, expiryDate.toString())
                .addToAnnotations(ACME_CA_ANNOTATION, certificate.ca)
                .addToAnnotations(REQUEST_ANNOTATION, domains)
                .endMetadata()
                .withData(certificate.certificateFiles)
                .done()

        log.info("Updated secret {} in namespace {}", secretName, namespace)
    }

    companion object {
        private val log = LoggerFactory.getLogger(this::class.java)
        private val gson = Gson()

        /**
         * Checks whether a certificate needs renewal. Renewal can be caused either by certificate expiry
         * or if the list of domains requested for a SAN certificate changes.
         */
        fun certificateNeedsRenewal(domains: List<String>, secret: Secret): Boolean {
            val isExpiring = certificateIsExpiring(secret)
            val domainsChanged = haveDomainsChanged(domains, secret)

            return isExpiring || domainsChanged
        }

        fun certificateIsExpiring(secret: Secret): Boolean {
            val expiryDate = getExpiryDate(secret)

            if (expiryDate.isDefined()) {
                return LocalDate.now().isAfter(expiryDate.get().minusDays(2))
            } else {
                log.warn("No expiry date set on secret {} in namespace {}!",
                        secret.metadata.name, secret.metadata.namespace)
                return false
            }
        }

        private fun getExpiryDate(secret: Secret): Option<LocalDate> {
            val annotation = secret.metadata.annotations[EXPIRY_ANNOTATION]
            return annotation.toOption().map { LocalDate.parse(it) }
        }

        private fun haveDomainsChanged(domains: List<String>, secret: Secret): Boolean {
            return secret.metadata.annotations[REQUEST_ANNOTATION].toOption()
                    .map {
                        val type = object : TypeToken<List<String>>() {}.type
                        val secretDomains: List<String> = gson.fromJson(it, type)
                        (secretDomains.size != domains.size) || !secretDomains.containsAll(domains)
                    }
                    .getOrElse {
                        log.warn("acme/certificate annotation missing on secret {}!", secret.metadata.name)
                        false
                    }
        }
    }
}